File: checkers.py
# checkers.py
import os
import copy
import random
import sys

SIZE = 8
EMPTY = 0
WHITE = 1
BLACK = 2
WHITE_KING = 3
BLACK_KING = 4

class Checkers:
    def __init__(self):
        self.board = [[EMPTY]*SIZE for _ in range(SIZE)]
        self.turn = WHITE
        self.selected = None
        self.valid_moves = []
        self.must_capture = False
        self.move_history = []
        self._init_board()
        self._find_all_captures()

    def _init_board(self):
        for row in range(3):
            for col in range(SIZE):
                if (row + col) % 2 == 1:
                    self.board[row][col] = BLACK
        for row in range(5, 8):
            for col in range(SIZE):
                if (row + col) % 2 == 1:
                    self.board[row][col] = WHITE

    def _is_king(self, piece):
        return piece in (WHITE_KING, BLACK_KING)

    def _is_white(self, piece):
        return piece in (WHITE, WHITE_KING)

    def _is_black(self, piece):
        return piece in (BLACK, BLACK_KING)

    def _opponent(self, piece):
        if self._is_white(piece):
            return BLACK
        return WHITE

    def _in_bounds(self, r, c):
        return 0 <= r < SIZE and 0 <= c < SIZE

    def _get_piece(self, r, c):
        return self.board[r][c] if self._in_bounds(r, c) else EMPTY

    def _is_enemy(self, r, c, piece):
        p = self._get_piece(r, c)
        if p == EMPTY:
            return False
        if self._is_white(piece):
            return self._is_black(p)
        else:
            return self._is_white(p)

    def _can_capture(self, r, c, dr, dc):
        nr, nc = r + dr, c + dc
        if not self._in_bounds(nr, nc):
            return False
        if self._get_piece(nr, nc) == EMPTY:
            return False
        if not self._is_enemy(nr, nc, self.board[r][c]):
            return False
        land_r, land_c = r + 2*dr, c + 2*dc
        if not self._in_bounds(land_r, land_c):
            return False
        if self._get_piece(land_r, land_c) != EMPTY:
            return False
        return True

    def _capture_moves_from(self, r, c, visited=None):
        if visited is None:
            visited = set()
        piece = self.board[r][c]
        dirs = [(-1,-1), (-1,1), (1,-1), (1,1)]
        if self._is_king(piece):
            # kings can move in all diagonal directions
            dirs = [(-1,-1), (-1,1), (1,-1), (1,1)]
        moves = []
        for dr, dc in dirs:
            if self._can_capture(r, c, dr, dc):
                land_r, land_c = r + 2*dr, c + 2*dc
                # Check that this position hasn't been visited in this chain
                if (land_r, land_c) in visited:
                    continue
                # Perform capture on a copy to check further captures
                new_board = [row[:] for row in self.board]
                captured = new_board[r+dr][c+dc]
                new_board[land_r][land_c] = piece
                new_board[r][c] = EMPTY
                new_board[r+dr][c+dc] = EMPTY
                # Check if the captured piece was a king
                # Not needed for move validation but for multi-jump we'll handle in move execution
                # Append the move sequence
                # For simplicity, we'll just record the move and continue recursively
                # We'll store the sequence in a list
                # We'll implement the recursive search in the move function, not here.
                moves.append((land_r, land_c))
        return moves

    def _find_all_captures_for_color(self, color):
        captures = []
        for r in range(SIZE):
            for c in range(SIZE):
                p = self.board[r][c]
                if p == EMPTY:
                    continue
                if self._is_white(p) and color != WHITE:
                    continue
                if self._is_black(p) and color != BLACK:
                    continue
                # Check capture moves
                dirs = [(-1,-1), (-1,1), (1,-1), (1,1)]
                if self._is_king(p):
                    # King can capture in all diagonal directions
                    pass
                for dr, dc in dirs:
                    if self._can_capture(r, c, dr, dc):
                        captures.append((r, c))
                        break
        return captures

    def _find_all_moves_for_piece(self, r, c):
        piece = self.board[r][c]
        moves = []
        # Check captures first
        dirs = [(-1,-1), (-1,1), (1,-1), (1,1)]
        if self._is_king(piece):
            # kings can move both directions
            pass
        # Simple moves (non-capture)
        for dr, dc in dirs:
            nr, nc = r + dr, c + dc
            if self._in_bounds(nr, nc) and self._get_piece(nr, nc) == EMPTY:
                # Check if move direction is forward for non-kings
                if not self._is_king(piece):
                    if self._is_white(piece) and dr != -1:
                        continue
                    if self._is_black(piece) and dr != 1:
                        continue
                moves.append((nr, nc))
        return moves

    def _has_captures(self, color):
        captures = self._find_all_captures_for_color(color)
        return len(captures) > 0

    def _get_legal_moves(self, r, c):
        piece = self.board[r][c]
        if piece == EMPTY:
            return []
        if self._is_white(piece) and self.turn != WHITE:
            return []
        if self._is_black(piece) and self.turn != BLACK:
            return []
        # If there are mandatory captures, only capture moves are allowed
        if self._has_captures(self.turn):
            # Only return capture moves for this piece
            return self._capture_moves_from(r, c)
        else:
            return self._find_all_moves_for_piece(r, c)

    def _is_winner(self):
        # Check if any player has no pieces left
        white_pieces = 0
        black_pieces = 0
        for r in range(SIZE):
            for c in range(SIZE):
                p = self.board[r][c]
                if self._is_white(p):
                    white_pieces += 1
                elif self._is_black(p):
                    black_pieces += 1
        if white_pieces == 0:
            return BLACK
        if black_pieces == 0:
            return WHITE
        return None

    def apply_move(self, from_r, from_c, to_r, to_c):
        piece = self.board[from_r][from_c]
        if piece == EMPTY:
            return False
        # Check if it's a capture
        dr = to_r - from_r
        dc = to_c - from_c
        if abs(dr) == 2 and abs(dc) == 2:
            # Capture move
            captured_r = from_r + dr//2
            captured_c = from_c + dc//2
            if self._get_piece(captured_r, captured_c) == EMPTY:
                return False
            # Remove captured piece
            self.board[captured_r][captured_c] = EMPTY
        # Move the piece
        self.board[to_r][to_c] = piece
        self.board[from_r][from_c] = EMPTY
        # Check promotion
        if not self._is_king(piece):
            if self._is_white(piece) and to_r == 0:
                self.board[to_r][to_c] = WHITE_KING
            elif self._is_black(piece) and to_r == SIZE-1:
                self.board[to_r][to_c] = BLACK_KING
        # Switch turn
        self.turn = self._opponent(piece)  # Actually opponent color
        # For simplicity, we'll just switch the turn color
        if self.turn == WHITE:
            self.turn = BLACK
        else:
            self.turn = WHITE
        return True

    def undo_move(self):
        # Not fully implemented for demo
        pass

    def display(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print("  a b c d e f g h")
        for r in range(SIZE):
            print(f"{r+1} ", end='')
            for c in range(SIZE):
                p = self.board[r][c]
                if p == EMPTY:
                    ch = '.'
                elif p == WHITE:
                    ch = 'w'
                elif p == BLACK:
                    ch = 'b'
                elif p == WHITE_KING:
                    ch = 'W'
                elif p == BLACK_KING:
                    ch = 'B'
                print(ch, end=' ')
            print(f"{r+1}")
        print("  a b c d e f g h")
        print(f"Turn: {'White' if self.turn == WHITE else 'Black'}")

    def parse_move(self, move_str):
        # Expect format like 'a3-b4'
        try:
            parts = move_str.split('-')
            if len(parts) != 2:
                return None
            from_sq = parts[0].strip()
            to_sq = parts[1].strip()
            from_col = ord(from_sq[0]) - ord('a')
            from_row = int(from_sq[1]) - 1
            to_col = ord(to_sq[0]) - ord('a')
            to_row = int(to_sq[1]) - 1
            if not (0 <= from_row < SIZE and 0 <= from_col < SIZE and
                    0 <= to_row < SIZE and 0 <= to_col < SIZE):
                return None
            return (from_row, from_col, to_row, to_col)
        except:
            return None

    def player_move(self):
        while True:
            self.display()
            move_str = input(f"{'White' if self.turn == WHITE else 'Black'}, enter move (e.g., a3-b4) or 'q' to quit: ").strip()
            if move_str.lower() == 'q':
                return False
            if move_str.lower() == 'u':
                # Undo not implemented in demo
                print("Undo not available.")
                continue
            parsed = self.parse_move(move_str)
            if parsed is None:
                print("Invalid format. Use e.g., a3-b4")
                continue
            fr, fc, tr, tc = parsed
            piece = self.board[fr][fc]
            if piece == EMPTY:
                print("No piece at source.")
                continue
            if self._is_white(piece) and self.turn != WHITE:
                print("It's not your turn.")
                continue
            if self._is_black(piece) and self.turn != BLACK:
                print("It's not your turn.")
                continue
            legal = self._get_legal_moves(fr, fc)
            if (tr, tc) not in legal:
                print("Illegal move.")
                continue
            # Apply move
            if not self.apply_move(fr, fc, tr, tc):
                print("Move failed.")
                continue
            # Check win
            winner = self._is_winner()
            if winner:
                self.display()
                print(f"{'White' if winner == WHITE else 'Black'} wins!")
                return False
            # If the capturing piece can continue capturing, the player must continue (multi-jump)
            # We'll handle this by checking if the piece that moved has more captures and if so, force the player to move again.
            # In our simple version, we'll just let the turn switch and the next player can move.
            # For simplicity, we'll skip multi-jump forcing for now.
            return True

    def ai_move(self):
        # Simple AI: find all legal moves and pick a random one
        moves = []
        for r in range(SIZE):
            for c in range(SIZE):
                p = self.board[r][c]
                if p == EMPTY:
                    continue
                if self._is_white(p) and self.turn != WHITE:
                    continue
                if self._is_black(p) and self.turn != BLACK:
                    continue
                legal = self._get_legal_moves(r, c)
                for tr, tc in legal:
                    moves.append((r, c, tr, tc))
        if not moves:
            print("No moves available.")
            return False
        # Random move
        move = random.choice(moves)
        fr, fc, tr, tc = move
        self.apply_move(fr, fc, tr, tc)
        winner = self._is_winner()
        if winner:
            self.display()
            print(f"{'White' if winner == WHITE else 'Black'} wins!")
            return False
        return True

def main():
    game = Checkers()
    print("Russian Checkers")
    print("Controls: enter moves as 'a3-b4'. 'q' to quit.")
    while True:
        if game.turn == WHITE:
            if not game.player_move():
                break
        else:
            if not game.ai_move():
                break
        # Check if game ended
        winner = game._is_winner()
        if winner:
            game.display()
            print(f"{'White' if winner == WHITE else 'Black'} wins!")
            break

if __name__ == "__main__":
    main()

File: checkers.rb
# checkers.rb
class Checkers
  SIZE = 8
  EMPTY = 0; WHITE = 1; BLACK = 2; WHITE_KING = 3; BLACK_KING = 4

  attr_reader :board, :turn

  def initialize
    @board = Array.new(SIZE) { Array.new(SIZE, EMPTY) }
    @turn = WHITE
    init_board
  end

  def init_board
    (0...3).each do |r|
      (0...SIZE).each do |c|
        @board[r][c] = BLACK if (r + c).odd?
      end
    end
    (5...8).each do |r|
      (0...SIZE).each do |c|
        @board[r][c] = WHITE if (r + c).odd?
      end
    end
  end

  def is_white(p) = [WHITE, WHITE_KING].include?(p)
  def is_black(p) = [BLACK, BLACK_KING].include?(p)
  def is_king(p) = [WHITE_KING, BLACK_KING].include?(p)
  def opponent(p) = is_white(p) ? BLACK : WHITE

  def in_bounds(r, c) = r >= 0 && r < SIZE && c >= 0 && c < SIZE
  def get_piece(r, c) = in_bounds(r, c) ? @board[r][c] : EMPTY

  def is_enemy(r, c, piece)
    p = get_piece(r, c)
    return false if p == EMPTY
    return true if is_white(piece) && is_black(p)
    return true if is_black(piece) && is_white(p)
    false
  end

  def can_capture(r, c, dr, dc)
    nr, nc = r + dr, c + dc
    return false unless in_bounds(nr, nc)
    return false if get_piece(nr, nc) == EMPTY
    return false unless is_enemy(nr, nc, @board[r][c])
    lr, lc = r + 2*dr, c + 2*dc
    return false unless in_bounds(lr, lc)
    return false if get_piece(lr, lc) != EMPTY
    true
  end

  def capture_moves_from(r, c)
    piece = @board[r][c]
    dirs = [[-1,-1], [-1,1], [1,-1], [1,1]]
    if !is_king(piece)
      if is_white(piece)
        dirs = [[-1,-1], [-1,1]]
      else
        dirs = [[1,-1], [1,1]]
      end
    end
    moves = []
    dirs.each do |dr, dc|
      if can_capture(r, c, dr, dc)
        moves << [r + 2*dr, c + 2*dc]
      end
    end
    moves
  end

  def simple_moves_from(r, c)
    piece = @board[r][c]
    dirs = [[-1,-1], [-1,1], [1,-1], [1,1]]
    if !is_king(piece)
      if is_white(piece)
        dirs = [[-1,-1], [-1,1]]
      else
        dirs = [[1,-1], [1,1]]
      end
    end
    moves = []
    dirs.each do |dr, dc|
      nr, nc = r + dr, c + dc
      if in_bounds(nr, nc) && get_piece(nr, nc) == EMPTY
        moves << [nr, nc]
      end
    end
    moves
  end

  def has_captures(color)
    (0...SIZE).each do |r|
      (0...SIZE).each do |c|
        p = @board[r][c]
        next if p == EMPTY
        next if color == WHITE && !is_white(p)
        next if color == BLACK && !is_black(p)
        return true if capture_moves_from(r, c).any?
      end
    end
    false
  end

  def legal_moves_from(r, c)
    piece = @board[r][c]
    return [] if piece == EMPTY
    return [] if is_white(piece) && @turn != WHITE
    return [] if is_black(piece) && @turn != BLACK
    if has_captures(@turn)
      capture_moves_from(r, c)
    else
      simple_moves_from(r, c)
    end
  end

  def apply_move(fr, fc, tr, tc)
    piece = @board[fr][fc]
    return false if piece == EMPTY
    dr, dc = tr - fr, tc - fc
    if dr.abs == 2 && dc.abs == 2
      cr, cc = fr + dr/2, fc + dc/2
      return false if get_piece(cr, cc) == EMPTY
      @board[cr][cc] = EMPTY
    end
    @board[tr][tc] = piece
    @board[fr][fc] = EMPTY
    if !is_king(piece)
      if is_white(piece) && tr == 0
        @board[tr][tc] = WHITE_KING
      elsif is_black(piece) && tr == SIZE-1
        @board[tr][tc] = BLACK_KING
      end
    end
    @turn = opponent(piece)
    true
  end

  def winner
    white = black = 0
    (0...SIZE).each do |r|
      (0...SIZE).each do |c|
        p = @board[r][c]
        white += 1 if is_white(p)
        black += 1 if is_black(p)
      end
    end
    return BLACK if white == 0
    return WHITE if black == 0
    EMPTY
  end

  def display
    system('clear') || system('cls')
    puts "  a b c d e f g h"
    (0...SIZE).each do |r|
      print "#{r+1} "
      (0...SIZE).each do |c|
        p = @board[r][c]
        ch = case p
             when EMPTY then '.'
             when WHITE then 'w'
             when BLACK then 'b'
             when WHITE_KING then 'W'
             when BLACK_KING then 'B'
             end
        print "#{ch} "
      end
      puts r+1
    end
    puts "  a b c d e f g h"
    puts "Turn: #{@turn == WHITE ? 'White' : 'Black'}"
  end

  def parse_move(str)
    parts = str.split('-')
    return nil if parts.length != 2
    from, to = parts[0].strip, parts[1].strip
    return nil if from.length != 2 || to.length != 2
    fc = from[0].ord - 'a'.ord
    fr = from[1].to_i - 1
    tc = to[0].ord - 'a'.ord
    tr = to[1].to_i - 1
    return nil if [fr, fc, tr, tc].any? { |v| v < 0 || v >= SIZE }
    [fr, fc, tr, tc]
  end

  def player_move
    loop do
      display
      color = @turn == WHITE ? 'White' : 'Black'
      print "#{color}, enter move (e.g., a3-b4) or 'q' to quit: "
      input = gets.chomp.strip
      return false if input == 'q'
      if input == 'u'
        puts "Undo not implemented."
        next
      end
      parsed = parse_move(input)
      unless parsed
        puts "Invalid format."
        next
      end
      fr, fc, tr, tc = parsed
      piece = @board[fr][fc]
      if piece == EMPTY
        puts "No piece there."
        next
      end
      if (is_white(piece) && @turn != WHITE) || (is_black(piece) && @turn != BLACK)
        puts "Not your turn."
        next
      end
      legal = legal_moves_from(fr, fc)
      unless legal.include?([tr, tc])
        puts "Illegal move."
        next
      end
      unless apply_move(fr, fc, tr, tc)
        puts "Move failed."
        next
      end
      w = winner
      if w != EMPTY
        display
        puts w == WHITE ? "White wins!" : "Black wins!"
        return false
      end
      return true
    end
  end

  def ai_move
    moves = []
    (0...SIZE).each do |r|
      (0...SIZE).each do |c|
        p = @board[r][c]
        next if p == EMPTY
        next if is_white(p) && @turn != WHITE
        next if is_black(p) && @turn != BLACK
        legal = legal_moves_from(r, c)
        legal.each { |tr, tc| moves << [r, c, tr, tc] }
      end
    end
    if moves.empty?
      puts "AI has no moves."
      return false
    end
    move = moves.sample
    apply_move(move[0], move[1], move[2], move[3])
    w = winner
    if w != EMPTY
      display
      puts w == WHITE ? "White wins!" : "Black wins!"
      return false
    end
    true
  end
end

game = Checkers.new
puts "Russian Checkers"
puts "Controls: enter moves as 'a3-b4'. 'q' to quit."
loop do
  if game.turn == Checkers::WHITE
    break unless game.player_move
  else
    break unless game.ai_move
  end
end
